{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np \n",
    "import pathlib as pl \n",
    "import requests\n",
    "import country_converter as coco \n",
    "from module import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# define path \n",
    "DIR = pl.Path(\"../..\")\n",
    "\n",
    "data = DIR / \"data\"\n",
    "internal = DIR / \"data\" / \"ersilia\"\n",
    "external = DIR / \"data\" / \"external\"\n",
    "PROC_DIR = DIR / \"src\" / \"processed\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# import internal data\n",
    "pub = pd.read_csv(internal / \"publications.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>slug</th>\n",
       "      <th>sdg</th>\n",
       "      <th>open</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chemocentric-view</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>similar-binding-pockets</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>oncogenomic-landscape</td>\n",
       "      <td>None</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rationalizing-drug-response</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              slug                         sdg   open\n",
       "0  structural-systems-pharmacology  Good health and well-being  False\n",
       "1                chemocentric-view  Good health and well-being   True\n",
       "2          similar-binding-pockets  Good health and well-being   True\n",
       "3            oncogenomic-landscape                        None   True\n",
       "4      rationalizing-drug-response  Good health and well-being  False"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the SDGs and Open Access status for Ersilia's publication from OpenAlex API\n",
    "pub_info = {}\n",
    "for slug, id in zip(pub.Slug,pub.OpenAlexID):\n",
    "    response = get_work(id)\n",
    "    sdg = response[\"sustainable_development_goals\"]\n",
    "    oa_status = response['open_access']['is_oa']\n",
    "    if len(sdg) == 0:\n",
    "        pub_info[slug] = (None, oa_status)\n",
    "    else:\n",
    "        pub_info[slug] = (sdg[0]['display_name'], oa_status)\n",
    "\n",
    "pub_info_df = pd.DataFrame(pub_info).T.reset_index().rename(columns={'index': 'slug', 0: 'sdg', 1: 'open'})\n",
    "pub_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>slug</th>\n",
       "      <th>sdg</th>\n",
       "      <th>open</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>Closed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chemocentric-view</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>Open</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>similar-binding-pockets</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>Open</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>oncogenomic-landscape</td>\n",
       "      <td>None</td>\n",
       "      <td>Open</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rationalizing-drug-response</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>Closed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              slug                         sdg    open\n",
       "0  structural-systems-pharmacology  Good health and well-being  Closed\n",
       "1                chemocentric-view  Good health and well-being    Open\n",
       "2          similar-binding-pockets  Good health and well-being    Open\n",
       "3            oncogenomic-landscape                        None    Open\n",
       "4      rationalizing-drug-response  Good health and well-being  Closed"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Map open accessibility \n",
    "open_map = {\n",
    "    True: \"Open\",\n",
    "    False: 'Closed'\n",
    "}\n",
    "pub_info_df['open'] = pub_info_df['open'].map(open_map)\n",
    "pub_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "pub_info_merge_ersilia = pub.merge(pub_info_df, how='left', left_on='Slug', right_on = 'slug').drop(columns=['slug','Abstract']).fillna(\"None\") #dropping Abstract since it's causing issues uploading to Looker\n",
    "#pub_info_merge_ersilia.to_csv(PROC_DIR/\"publication_sdg_open.csv\", index=False)\n",
    "pub_info_merge_ersilia.to_csv(PROC_DIR/\"publication_merged.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get required information on the citations\n",
    "# information contains: citation name (cited_in_paper), SDG, SDG score, Publication Date, Primary Topic\n",
    "\n",
    "# Initialising\n",
    "citation_info = {}\n",
    "author_info = {}\n",
    "for slug, id in zip(pub.Slug,pub.OpenAlexID):\n",
    "    # Request and extract info from response from OpenAlex API\n",
    "    work_cited_url = get_cited_url(id)\n",
    "    response = get_cited(work_cited_url)\n",
    "    info = get_citation_info(response)\n",
    "    authors = get_author_info(response)\n",
    "    # Transform response into DataFrame\n",
    "    df1 = pd.DataFrame(info).T.reset_index().rename(columns={'index': 'cited_in_paper', 0: 'sdg_goal', 1:'score', 2:'publication_date', 3:'primary_topic'})\n",
    "    df1['Slug'] = [slug] * len(df1)\n",
    "\n",
    "    df2 = pd.DataFrame(authors).T.reset_index().rename(columns={'index': 'cited_in_paper', 0:'Institution', 1:'Country', 2: 'Author Name', 3: 'Author ID'})\n",
    "    try:\n",
    "        df2['cited_in_paper'] = df2['cited_in_paper'].str.split(\"_\").str[0]\n",
    "    except: \n",
    "        pass\n",
    "    df2['Slug'] = [slug] * len(df2)\n",
    "\n",
    "    # Store in dictionary\n",
    "    citation_info[slug] = df1\n",
    "    author_info[slug] = df2\n",
    "\n",
    "# Concat all dataframes in the dictionary for citation info\n",
    "citation_info_df = pd.concat(citation_info.values(), axis=0).reset_index(drop=True).rename(columns={'Slug':'slug'})\n",
    "citation_info_df['publication_date'] = pd.to_datetime(citation_info_df['publication_date'])\n",
    "citation_info_df['sdg_goal'] = citation_info_df['sdg_goal'].astype(str)\n",
    "\n",
    "# Concat all dataframes in the dictionary for authors info\n",
    "author_df = pd.concat(author_info.values(), axis=0).reset_index(drop=True).rename(columns={'Slug':'slug'})\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cited_in_paper</th>\n",
       "      <th>sdg_goal</th>\n",
       "      <th>score</th>\n",
       "      <th>publication_date</th>\n",
       "      <th>primary_topic</th>\n",
       "      <th>slug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>In silico methods to address polypharmacology:...</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>0.45</td>\n",
       "      <td>2016-02-01</td>\n",
       "      <td>Computational Methods in Drug Discovery</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2023-02-28</td>\n",
       "      <td>Protein Structure Prediction and Analysis</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Towards a detailed atlas of protein–protein in...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2013-12-01</td>\n",
       "      <td>Analysis of Gene Interaction Networks</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Alternative modulation of protein–protein inte...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2015-12-01</td>\n",
       "      <td>Targeted Protein Degradation in Biomedical Res...</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Machine Learning Toxicity Prediction: Latest A...</td>\n",
       "      <td>Good health and well-being</td>\n",
       "      <td>0.67</td>\n",
       "      <td>2022-12-13</td>\n",
       "      <td>Computational Methods in Drug Discovery</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      cited_in_paper  \\\n",
       "0  In silico methods to address polypharmacology:...   \n",
       "1  Before and after AlphaFold2: An overview of pr...   \n",
       "2  Towards a detailed atlas of protein–protein in...   \n",
       "3  Alternative modulation of protein–protein inte...   \n",
       "4  Machine Learning Toxicity Prediction: Latest A...   \n",
       "\n",
       "                     sdg_goal score publication_date  \\\n",
       "0  Good health and well-being  0.45       2016-02-01   \n",
       "1                        None  None       2023-02-28   \n",
       "2                        None  None       2013-12-01   \n",
       "3                        None  None       2015-12-01   \n",
       "4  Good health and well-being  0.67       2022-12-13   \n",
       "\n",
       "                                       primary_topic  \\\n",
       "0            Computational Methods in Drug Discovery   \n",
       "1          Protein Structure Prediction and Analysis   \n",
       "2              Analysis of Gene Interaction Networks   \n",
       "3  Targeted Protein Degradation in Biomedical Res...   \n",
       "4            Computational Methods in Drug Discovery   \n",
       "\n",
       "                              slug  \n",
       "0  structural-systems-pharmacology  \n",
       "1  structural-systems-pharmacology  \n",
       "2  structural-systems-pharmacology  \n",
       "3  structural-systems-pharmacology  \n",
       "4  structural-systems-pharmacology  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "citation_info_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cited_in_paper</th>\n",
       "      <th>Institution</th>\n",
       "      <th>Country</th>\n",
       "      <th>Author Name</th>\n",
       "      <th>Author ID</th>\n",
       "      <th>slug</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>In silico methods to address polypharmacology:...</td>\n",
       "      <td>None</td>\n",
       "      <td>[IT]</td>\n",
       "      <td>Antonio Lavecchia</td>\n",
       "      <td>https://openalex.org/A5083327774</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>In silico methods to address polypharmacology:...</td>\n",
       "      <td>None</td>\n",
       "      <td>[IT]</td>\n",
       "      <td>Carmen Cerchia</td>\n",
       "      <td>https://openalex.org/A5031786233</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>Letícia M. F. Bertoline</td>\n",
       "      <td>https://openalex.org/A5060313675</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>Angélica N. Lima</td>\n",
       "      <td>https://openalex.org/A5048517462</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>José Eduardo Krieger</td>\n",
       "      <td>https://openalex.org/A5069859243</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                      cited_in_paper  \\\n",
       "0  In silico methods to address polypharmacology:...   \n",
       "1  In silico methods to address polypharmacology:...   \n",
       "2  Before and after AlphaFold2: An overview of pr...   \n",
       "3  Before and after AlphaFold2: An overview of pr...   \n",
       "4  Before and after AlphaFold2: An overview of pr...   \n",
       "\n",
       "                 Institution Country              Author Name  \\\n",
       "0                       None    [IT]        Antonio Lavecchia   \n",
       "1                       None    [IT]           Carmen Cerchia   \n",
       "2  Universidade de São Paulo      BR  Letícia M. F. Bertoline   \n",
       "3  Universidade de São Paulo      BR         Angélica N. Lima   \n",
       "4  Universidade de São Paulo      BR     José Eduardo Krieger   \n",
       "\n",
       "                          Author ID                             slug  \n",
       "0  https://openalex.org/A5083327774  structural-systems-pharmacology  \n",
       "1  https://openalex.org/A5031786233  structural-systems-pharmacology  \n",
       "2  https://openalex.org/A5060313675  structural-systems-pharmacology  \n",
       "3  https://openalex.org/A5048517462  structural-systems-pharmacology  \n",
       "4  https://openalex.org/A5069859243  structural-systems-pharmacology  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "author_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save as csv file\n",
    "citation_info_df.drop(columns=['score']).to_csv(PROC_DIR/\"citation_information.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "((pd.merge(left = citation_info_df.groupby(citation_info_df.publication_date.dt.year).agg('count')['slug'].reset_index(), \n",
    "          right = pub_info_merge_ersilia.groupby(pub_info_merge_ersilia.Year).agg('count')['Slug'].reset_index(), \n",
    "          how='outer', \n",
    "          left_on='publication_date', \n",
    "          right_on='Year')\n",
    "          .drop(columns=['Year'])\n",
    "          .fillna(0)\n",
    "          .rename(columns={'publication_date': 'Publication Year',\n",
    "                           'slug': 'Citation',\n",
    "                           'Slug': 'Publication'}))\n",
    "                           .melt(id_vars='Publication Year')\n",
    "                           .rename(columns={\n",
    "                               'variable': 'Type',\n",
    "                               'value': 'Count'\n",
    "                           }).to_csv(PROC_DIR/\"summary_of_publication_and_citation_by_year.csv\", index=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n",
      "None not found in regex\n"
     ]
    }
   ],
   "source": [
    "# Mapping the country code to country short name\n",
    "author_flatten_df = author_df.copy().explode(['Country']).fillna(value = {'Country': \"None\"}).reset_index(drop=True)\n",
    "author_flatten_df['short_name'] = author_flatten_df.Country.apply(lambda x: coco.convert(names = x, to='name_short', not_found=None))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "country_to_region_mapping = pd.read_csv(external/\"geo\"/\"country_mapping.csv\")\n",
    "country_to_region_mapping.head()\n",
    "\n",
    "region_mapping = {\n",
    "    \"GS\" : \"Global South\",\n",
    "    \"GN\" : \"Global North\"\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cited_in_paper</th>\n",
       "      <th>Institution</th>\n",
       "      <th>Country</th>\n",
       "      <th>Author Name</th>\n",
       "      <th>Author ID</th>\n",
       "      <th>slug</th>\n",
       "      <th>short_name</th>\n",
       "      <th>Regions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>In silico methods to address polypharmacology:...</td>\n",
       "      <td>None</td>\n",
       "      <td>IT</td>\n",
       "      <td>Antonio Lavecchia</td>\n",
       "      <td>https://openalex.org/A5083327774</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>In silico methods to address polypharmacology:...</td>\n",
       "      <td>None</td>\n",
       "      <td>IT</td>\n",
       "      <td>Carmen Cerchia</td>\n",
       "      <td>https://openalex.org/A5031786233</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Italy</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>Letícia M. F. Bertoline</td>\n",
       "      <td>https://openalex.org/A5060313675</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Global South</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>Angélica N. Lima</td>\n",
       "      <td>https://openalex.org/A5048517462</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Global South</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Before and after AlphaFold2: An overview of pr...</td>\n",
       "      <td>Universidade de São Paulo</td>\n",
       "      <td>BR</td>\n",
       "      <td>José Eduardo Krieger</td>\n",
       "      <td>https://openalex.org/A5069859243</td>\n",
       "      <td>structural-systems-pharmacology</td>\n",
       "      <td>Brazil</td>\n",
       "      <td>Global South</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5872</th>\n",
       "      <td>Worldwide study of the taste of bitter medicin...</td>\n",
       "      <td>Monell Chemical Senses Center</td>\n",
       "      <td>US</td>\n",
       "      <td>Riley D Herriman</td>\n",
       "      <td>https://openalex.org/A5087531977</td>\n",
       "      <td>ai-tailor-drugs-africa-comment</td>\n",
       "      <td>United States</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5873</th>\n",
       "      <td>Worldwide study of the taste of bitter medicin...</td>\n",
       "      <td>National Institute on Alcohol Abuse and Alcoho...</td>\n",
       "      <td>US</td>\n",
       "      <td>Paule V. Joseph</td>\n",
       "      <td>https://openalex.org/A5058203848</td>\n",
       "      <td>ai-tailor-drugs-africa-comment</td>\n",
       "      <td>United States</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5874</th>\n",
       "      <td>Worldwide study of the taste of bitter medicin...</td>\n",
       "      <td>Countess of Chester Hospital</td>\n",
       "      <td>GB</td>\n",
       "      <td>Oghogho Braimah</td>\n",
       "      <td>https://openalex.org/A5095921631</td>\n",
       "      <td>ai-tailor-drugs-africa-comment</td>\n",
       "      <td>United Kingdom</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5875</th>\n",
       "      <td>Worldwide study of the taste of bitter medicin...</td>\n",
       "      <td>Monell Chemical Senses Center</td>\n",
       "      <td>US</td>\n",
       "      <td>Danielle R. Reed</td>\n",
       "      <td>https://openalex.org/A5074936451</td>\n",
       "      <td>ai-tailor-drugs-africa-comment</td>\n",
       "      <td>United States</td>\n",
       "      <td>Global North</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5876</th>\n",
       "      <td>Predicting ligand–protein interactions</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Sarah Crunkhorn</td>\n",
       "      <td>https://openalex.org/A5020125353</td>\n",
       "      <td>chemoproteomics-cemm</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5877 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                         cited_in_paper  \\\n",
       "0     In silico methods to address polypharmacology:...   \n",
       "1     In silico methods to address polypharmacology:...   \n",
       "2     Before and after AlphaFold2: An overview of pr...   \n",
       "3     Before and after AlphaFold2: An overview of pr...   \n",
       "4     Before and after AlphaFold2: An overview of pr...   \n",
       "...                                                 ...   \n",
       "5872  Worldwide study of the taste of bitter medicin...   \n",
       "5873  Worldwide study of the taste of bitter medicin...   \n",
       "5874  Worldwide study of the taste of bitter medicin...   \n",
       "5875  Worldwide study of the taste of bitter medicin...   \n",
       "5876             Predicting ligand–protein interactions   \n",
       "\n",
       "                                            Institution Country  \\\n",
       "0                                                  None      IT   \n",
       "1                                                  None      IT   \n",
       "2                             Universidade de São Paulo      BR   \n",
       "3                             Universidade de São Paulo      BR   \n",
       "4                             Universidade de São Paulo      BR   \n",
       "...                                                 ...     ...   \n",
       "5872                      Monell Chemical Senses Center      US   \n",
       "5873  National Institute on Alcohol Abuse and Alcoho...      US   \n",
       "5874                       Countess of Chester Hospital      GB   \n",
       "5875                      Monell Chemical Senses Center      US   \n",
       "5876                                               None    None   \n",
       "\n",
       "                  Author Name                         Author ID  \\\n",
       "0           Antonio Lavecchia  https://openalex.org/A5083327774   \n",
       "1              Carmen Cerchia  https://openalex.org/A5031786233   \n",
       "2     Letícia M. F. Bertoline  https://openalex.org/A5060313675   \n",
       "3            Angélica N. Lima  https://openalex.org/A5048517462   \n",
       "4        José Eduardo Krieger  https://openalex.org/A5069859243   \n",
       "...                       ...                               ...   \n",
       "5872         Riley D Herriman  https://openalex.org/A5087531977   \n",
       "5873          Paule V. Joseph  https://openalex.org/A5058203848   \n",
       "5874          Oghogho Braimah  https://openalex.org/A5095921631   \n",
       "5875         Danielle R. Reed  https://openalex.org/A5074936451   \n",
       "5876          Sarah Crunkhorn  https://openalex.org/A5020125353   \n",
       "\n",
       "                                 slug      short_name       Regions  \n",
       "0     structural-systems-pharmacology           Italy  Global North  \n",
       "1     structural-systems-pharmacology           Italy  Global North  \n",
       "2     structural-systems-pharmacology          Brazil  Global South  \n",
       "3     structural-systems-pharmacology          Brazil  Global South  \n",
       "4     structural-systems-pharmacology          Brazil  Global South  \n",
       "...                               ...             ...           ...  \n",
       "5872   ai-tailor-drugs-africa-comment   United States  Global North  \n",
       "5873   ai-tailor-drugs-africa-comment   United States  Global North  \n",
       "5874   ai-tailor-drugs-africa-comment  United Kingdom  Global North  \n",
       "5875   ai-tailor-drugs-africa-comment   United States  Global North  \n",
       "5876             chemoproteomics-cemm            None           NaN  \n",
       "\n",
       "[5877 rows x 8 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "author_flatten_with_region_df = author_flatten_df.merge(country_to_region_mapping, how='left', left_on=\"short_name\", right_on=\"Countries\").drop(columns='Countries')\n",
    "author_flatten_with_region_df['Global North/ Global South classification'] = author_flatten_with_region_df['Global North/ Global South classification'].map(region_mapping)\n",
    "author_flatten_with_region_df.rename(columns={\"Global North/ Global South classification\":\"Regions\"}, inplace=True)\n",
    "author_flatten_with_region_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge with citation information to get publication and topic\n",
    "author_flatten_with_region_df.merge(citation_info_df[['cited_in_paper', 'publication_date', 'primary_topic']].drop_duplicates(subset=['cited_in_paper'], keep=\"first\"), how=\"left\", left_on=\"cited_in_paper\", right_on=\"cited_in_paper\").sort_values(by='publication_date', ascending=True).drop_duplicates(subset='Author ID', keep='first').to_csv(PROC_DIR/\"author_information.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
